### PR TITLE
fix: #528 - Preventing duplicate server names within the same project

### PR DESCRIPTION
**Pull request by nathanielcompton**
Refactoring Serializers, viewsets, and request methods for server project to ensure that multiple servers all within the same 3Blades project have unique server names

### CODE REVIEW

File: servers/models.py [hunk 1]
```diff
@@ -58,6 +58,8 @@ class Server(models.Model):
     last_start = models.DateTimeField(null=True)
 
     class Meta:
+        # Ensuring Projects will not have duplicate Server names
```
Comment (raygesualdo) [inline servers/models.py]: A total nitpick :smile: but this should probably read "Ensure projects will not have..."

Comment (nathanielcompton) [reply servers/models.py]: Fixed.

File: servers/views.py [hunk 2]
```diff
@@ -21,18 +23,79 @@
...
+    return project
+
 class ServerViewSet(LookupByMultipleFields, viewsets.ModelViewSet):
     queryset = models.Server.objects.all()
     serializer_class = serializers.ServerSerializer
     permission_classes = (IsAuthenticated, ProjectChildPermission, TeamGroupPermission)
     filter_fields = ("name",)
     lookup_url_kwarg = 'server'
 
+    def _update(self, request, partial, *args, **kwargs):
```
Comment (raygesualdo) [inline servers/views.py]: Should `partial` be used somewhere in this function?

File: servers/models.py [hunk 3]
```diff
@@ -58,6 +58,8 @@ class Server(models.Model):
     last_start = models.DateTimeField(null=True)
 
     class Meta:
+        # Ensure Projects will not have duplicate Server names
+        unique_together = (('project', 'name'),)
```
Comment (johngriebel) [inline servers/models.py]: There should be a migration to go along with this. Run `python manage.py makemigrations` to generate it.

File: servers/tests/test_views.py [hunk 4]
```diff
@@ -231,6 +231,19 @@ def setUp(self):
         token = create_auth_jwt(self.user)
         self.client = self.client_class(HTTP_AUTHORIZATION=f'Bearer {token}')
 
+    def test_validate_server_name_prevents_duplicate_name(self):
+        # Passing a project server name identical to an existing server name should error
+        server = ServerFactory(project=self.project, name='test_name')
```
Comment (johngriebel) [inline servers/tests/test_views.py]: This is picking nits; but we should let factory boy generate the server's name, and simply pass `server.name` into the `data` dictionary you create later. This helps us validate that a.) our system can handle a wide range of characters, and b.) the maximum lengths of our char fields are consistently valid. We can discuss b a bit more in person so I can give a concrete example of what I mean.

File: servers/views.py [hunk 5]
```diff
@@ -1,5 +1,6 @@
 import logging
-from django.db.models import Sum, Max, F
+from django.conf import settings
+from django.db.models import Sum, Max, F, Q
```
Comment (johngriebel) [inline servers/views.py]: I don't think you ended up using`Q`, so this import can be removed.

File: servers/views.py [hunk 6]
```diff
@@ -21,18 +23,79 @@
...
+    def _update(self, request, partial, *args, **kwargs):
+        project = check_project_details_before_lookup(self, request, **kwargs)
+        servers = Server.objects.filter(project=project)
+        server = servers.tbs_filter(kwargs['server']).first()
+
+        data = request.data
+        if data.get('name') == server.name:
+            data.pop('name')
+
+        # Pass in required context keys
```
Comment (johngriebel) [inline servers/views.py]: `partial` should be passed into the instantiation of the serializer here. Although it's strange that none of the `partial_update` tests fail without it being passed. ¯\_(ツ)_/¯

File: servers/views.py [hunk 7]
```diff
@@ -21,18 +23,79 @@
...
+        serializer = self.get_serializer_class()(data=request.data,
+                                                 context={
+                                                     'project': project,
+                                                     'request': request,
+                                                     'version': self.kwargs.get('version', settings.DEFAULT_VERSION)
+                                                 })
+        serializer.is_valid(raise_exception=True)
+        self.perform_create(serializer)
+        return Response(serializer.data, status=status.HTTP_201_CREATED)
+        server = serializer.instance
```
Comment (johngriebel) [inline servers/views.py]: This line can't be reached. It must be a copy/paste issue or something like that.

Comment (nathanielcompton) [reply servers/views.py]: Updated with correction

File: servers/views.py (hunk 2 above)
Comment (nathanielcompton) [reply servers/views.py]: Updated with correction

File: servers/views.py (hunk 5 above)
Comment (nathanielcompton) [reply servers/views.py]: Updated with correction

File: servers/models.py (hunk 3 above)
Comment (nathanielcompton) [reply servers/models.py]: Updated with correction

File: servers/views.py (hunk 6 above)
Comment (nathanielcompton) [reply servers/views.py]: Updated with correction

**Pull request merged**